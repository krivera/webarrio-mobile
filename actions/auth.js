export const AUTH_TOKEN_UPDATED = "AUTH_TOKEN_UPDATED";
export const DESTROY_TOKEN = "DESTROY_TOKEN";

export function updateToken(authToken){
  return{
    authToken: authToken,
    type: AUTH_TOKEN_UPDATED
  }
}

export function closeSession(){
  return {
    type: DESTROY_TOKEN
  }
}