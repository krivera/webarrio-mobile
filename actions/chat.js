import firebase from "../api/firebase";

export const CHATS_LOADED = 'CHATS_LOADED';
export const ADD_CHAT = 'ADD_CHAT';
export const ADD_MESSAGES_TO_CHAT = 'ADD_MESSAGES_TO_CHAT';
export const REQUEST_CHATS = 'REQUEST_CHATS';

export function receiveChats (rawChats){
  let chats = Object.keys(rawChats || {}).map(
    userId => ({...rawChats[userId], userId: userId})
  ).sort((a,b) => (b.updatedAt || b.createdAt) - (a.updatedAt || a.createdAt));
  return {
    type: CHATS_LOADED,
    chats: chats
  }
}

export function requestChats(){
  return {
    type: REQUEST_CHATS
  }
}

function receiveMessages (chatId, messages) {
  return {
    type: ADD_MESSAGES_TO_CHAT,
    chatId: chatId,
    messages: messages
  }
}

export function fetchMessages (chatId) {
  return dispatch => {
    firebase.database().ref().child(`/messages/${chatId}`)
    .on(
      'value',
      snapshot => dispatch(receiveMessages(chatId, snapshot.val()))
    )
  }
}

export function sendMessage (chatId, message, neighborhood, sender, receiver) {
  const ref = firebase.database().ref();
  const newMsgRef = ref.child(`/messages/${chatId}/${message._id}`)
    .set(message);
  ref.child(`/users/${sender}/chats/${neighborhood}/${receiver}`).update({
    lastMessage: message.text,
    updatedAt: message.createdAt,
  });
  ref.child(`/users/${receiver}/chats/${neighborhood}/${sender}`).update({
    lastMessage: message.text,
    updatedAt: message.createdAt,
  });
  let messages = {};
  messages[message._id] = message
  return {
    type: ADD_MESSAGES_TO_CHAT,
    chatId,
    messages
  }
}
