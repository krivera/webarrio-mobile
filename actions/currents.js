import { setNeighborhoodTree } from './neighborhood'

export const SET_CURRENT = 'SET_CURRENT'

export function setCurrent(key, value) {
  return {
    type: SET_CURRENT,
    key: key,
    value: value
  }
}

export function setInitialData(data) {
  return (dispatch) => {
    dispatch(setCurrent('user', data.user))
    dispatch(setNeighborhoodTree(data.neighborhoods))
    if (data.neighborhoods && data.neighborhoods.length) {
      let neighborhood = data.neighborhoods[0]
      dispatch(setCurrent('neighborhood', neighborhood))
      if (neighborhood.neighborhood_units && neighborhood.neighborhood_units.length) {
        let unit = neighborhood.neighborhood_units[0]
        dispatch(setCurrent('unit', unit))
        if (unit.apartments && unit.apartments.length) {
          dispatch(setCurrent('apartment', unit.apartments[0]))
        }
      }
    }
  }
}
