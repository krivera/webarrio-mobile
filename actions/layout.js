export const SET_TABBAR_HEIGHT = 'SET_TABBAR_HEIGHT'

export const setTabbarHeight = height => ({
  type: SET_TABBAR_HEIGHT,
  height
})
