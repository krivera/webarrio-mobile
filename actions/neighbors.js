import Axios from 'axios'
import { API_URL } from 'react-native-dotenv'

export const NEIGHBORS_LOADED = 'NEIGHBORS_LOADED'
export const REQUESTING = 'REQUESTING'

function isRequesting() {
  return {
    type: REQUESTING
  }
}

function receiveNeighbors(neighbors) {
  return {
    type: NEIGHBORS_LOADED,
    neighbors: neighbors
  }
}

export function fetchNeighbors(neighborhood, authToken) {
  return dispatch => {
    dispatch(isRequesting())
    Axios.get(`${API_URL}/neighborhoods/${neighborhood}/neighbors`,
      {
        headers: {
          Authorization: authToken
        }
      }
    ).then(response => {
      dispatch(receiveNeighbors(response.data.neighbors))
    })
  }
}
