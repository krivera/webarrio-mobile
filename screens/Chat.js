import React from 'react';
import{
  Dimensions,
  KeyboardAvoidingView,
  Platform,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import { GiftedChat, Bubble, Send } from 'react-native-gifted-chat';
import { Ionicons } from '@expo/vector-icons';
import { connect } from 'react-redux';
import { Header } from 'react-navigation';
import firebase from '../api/firebase';
import Colors from '../constants/Colors';
import BackButton from '../components/BackButton';
import { fetchMessages, sendMessage } from '../actions/chat';

const MESSAGE_ZERO = name => ({
  _id: 0,
  text: `Este es el comienzon de tu conversación con ${name}`,
  createdAt: new Date(),
  system: true
});

const { height: fullHeight } = Dimensions.get('window');

class ChatScreen extends React.Component{
  static navigationOptions = ({ navigation }) => ({
    headerTitle: 'Conversación',
    headerLeft: (<BackButton navigation={navigation} />),
    tabBarVisible: false
  });

  constructor(props){
    super(props);
    const { params } = props.navigation.state;
    this.chatId = params.chatId || (props.chats.find(
      chat =>
        chat.userId.toString() === params.user.id.toString()
      ) || { chatId: null }).chatId;
    this.state = {
      offset: 0
    };
  }

  componentWillMount(){
    if(this.chatId){
      this.props.dispatch(fetchMessages(this.chatId));
    }
  }

  onLayout = ({
    nativeEvent: { layout: { height } }
  }) => {
    const offset = fullHeight - height;
    this.setState({ offset });
  }

  send = (messages) => {
    const { user } = this.props.navigation.state.params;
    const {
      currentUser,
      currentNeighborhood,
      currentApartment,
      dispatch,
    } = this.props;
    if(!this.chatId){
      const createdAt = new Date();
      const ref = firebase.database().ref();
      const newRef = ref.child('/messages').push();
      this.chatId = newRef.key;
      const chatRef1 = ref.child(`/users/${currentUser.id}/chats/${currentNeighborhood.id}/${user.id}`).set({
        createdAt: createdAt,
        personPhoto: user.avatar,
        personName: user.name,
        personApartment: user.apartments[0].number,
        chatId: this.chatId
      });
      const chatRef2 = ref.child(`/users/${user.id}/chats/${currentNeighborhood.id}/${currentUser.id}`).set({
        createdAt: createdAt,
        personPhoto: currentUser.avatar,
        personName: currentUser.name,
        personApartment: currentApartment.number,
        chatId: this.chatId
      });
    }

    for(let message of messages){
      dispatch(sendMessage(
        this.chatId,
        { ...message, createdAt: message.createdAt.getTime() },
        currentNeighborhood.id,
        currentUser.id,
        user.id
      ));
    }
  }

  renderSend = (props) => (
    <Send {...props}>
      <View style={styles.sendButton}>
        <Text style={styles.sendText}>Enviar</Text>
      </View>
    </Send>);

  render(){
    const { currentUser, messages, currentApartment } = this.props;
    const chatUser = {
      _id: currentUser.id,
      avatar: currentUser.avatar,
      name: currentUser.name,
      apartment: currentApartment.number
    }
    const { user } = this.props.navigation.state.params;
    let this_messages = this.chatId
      ? Object.values(messages[this.chatId]).sort((a, b) => b.createdAt - a.createdAt)
      : [MESSAGE_ZERO(user.name)];
    return (
      <View style={styles.screen} onLayout={this.onLayout}>
        <GiftedChat
          messages={this_messages}
          onSend={this.send}
          user={chatUser}
          renderSend={this.renderSend}
          placeholder="Mensaje..."
          listViewProps={{
            style: {
              backgroundColor: 'white',
            },
          }}
        />
        {Platform.OS === 'android' && (
          <KeyboardAvoidingView
            behavior="padding"
            keyboardVerticalOffset={this.state.offset}
          />
        )}
      </View>
    );
  }
}

const mapStateToProps = state => ({
  currentUser: state.currentsReducer.user,
  currentNeighborhood: state.currentsReducer.neighborhood,
  currentApartment: state.currentsReducer.apartment,
  messages: state.chatReducer.messages,
  chats: state.chatReducer.chats
});

export default connect(mapStateToProps)(ChatScreen);

const styles = StyleSheet.create({
  sendText: {
    color: Colors.tintColor,
    fontSize: 18,
    fontWeight: 'bold'
  },
  sendButton: {
    alignSelf: 'center',
    padding: 5
  },
  screen: {
    flex: 1
  },
});
