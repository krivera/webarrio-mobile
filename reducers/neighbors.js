import { NEIGHBORS_LOADED, REQUESTING } from '../actions/neighbors';

const initialState = {
  neighbors: [],
  isRequesting: false
};

export default (state=initialState, action) => {
  switch(action.type){
    case NEIGHBORS_LOADED:
      return {
        neighbors: action.neighbors,
        isRequesting: false
      }
    case REQUESTING:
      return {
        neighbors: state.neighbors,
        isRequesting: true
      }
    default:
      return state;
  }
};
