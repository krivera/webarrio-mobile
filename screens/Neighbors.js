import React from 'react';
import { connect } from 'react-redux';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  View,
  Text,
  TextInput,
  TouchableOpacity
} from 'react-native';
import Avatar from '../components/Avatar';
import BackButton from '../components/BackButton';
import { fetchNeighbors } from '../actions/neighbors';
import Colors from '../constants/Colors';
import Ionicons from '@expo/vector-icons/Ionicons';

class NeighborsScreen extends React.Component{
  static navigationOptions = ({navigation}) => ({
    headerTitle: 'Vecinos',
    headerLeft: (<BackButton navigation={navigation} />)
  });

  constructor(props){
    super(props);
    this.state = {
      search: ''
    }
  }

  componentWillMount(){
    const { authToken, dispatch, currentNeighborhood } = this.props;
    dispatch(fetchNeighbors(currentNeighborhood.id, authToken));
  }

  renderNeighbor = ({ item: neighbor }) => {
    const { navigation } = this.props;
    return (
      <TouchableOpacity
        style={styles.neighbor}
        onPress={() => navigation.navigate('Chat', {user: neighbor})}
      >
        <Avatar source={{uri: neighbor.avatar}} name={neighbor.name} />
        <Text>
          {neighbor.name} de {neighbor.apartments[0].number}
        </Text>
      </TouchableOpacity>
    )
  }

  render(){
    const { isLoading, neighbors } = this.props;
    let filteredNeighbors = neighbors.filter(
      n => {
        let name = n.name + ' ' + n.apartments.map(a => a.number).join(' ');
        return new RegExp(this.state.search, 'i').test(name);
      }
    );
    return(
      <View style={styles.screen}>
        <View style={styles.searchBar}>
          <Ionicons name="ios-search" size={20} color="#c7c7cd" />
          <TextInput
            placeholder="Buscar"
            value={this.state.search}
            onChangeText={t => this.setState({search: t})}
            underlineColorAndroid="transparent"
            style={styles.searchInput}
          />
        </View>
        {isLoading && (
          <View><ActivityIndicator /></View>
        )}
        <FlatList
          data={filteredNeighbors}
          renderItem={this.renderNeighbor}
          keyExtractor={(item, index) => `${index}`}
        />
      </View>
    )
  }
}

const mapStateToProps = state => ({
  neighbors: state.neighborsReducer.neighbors,
  isLoading: state.neighborsReducer.isRequesting,
  authToken: state.authReducer.authToken,
  currentNeighborhood: state.currentsReducer.neighborhood
});

export default connect(mapStateToProps)(NeighborsScreen);

const styles = StyleSheet.create({
  neighbor: {
    backgroundColor: 'white',
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: Colors.inputBorder,
    padding: 10,
    flexDirection: 'row',
    alignItems: 'center'
  },
  searchBar: {
    flexDirection: 'row',
    padding: 10,
    backgroundColor: 'white',
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: Colors.inputBorder,
    alignItems: 'center',
  },
  searchInput: {
    paddingHorizontal: 10,
    flex: 1
  },
  screen: {
    flex: 1,
  }
})