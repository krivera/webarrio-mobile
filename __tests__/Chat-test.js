/* global
  it
  beforeEach
  jest
  expect
*/
import 'react-native'
import React from 'react'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import sinon from 'sinon'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { mockActionCreators, createDispatchMockImplementation } from 'jest-mock-action-creators'
import Neighbors from '../screens/Neighbors'
import * as neighborsActions from '../actions/neighbors'
import Chat from '../screens/Chat'

const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)
const initialState = {
  currentsReducer: {
    neighborhood: {
      id: 1,
      name: 'Test',
      address: 'test street 123'
    },
    user: {
      id: 2,
      name: 'current test',
      apartments: [
        {
          id: 2,
          number: '2'
        }
      ]
    },
    apartment: {
      number: '2',
      id: 2
    }
  },
  neighborsReducer: {
    neighbors: [
      {
        id: 1,
        name: 'test',
        last_name: 'neighbor',
        apartments: [
          {
            number: '1',
            id: 1
          }
        ]
      }
    ],
    isRequesting: false
  },
  authReducer: {
    authToken: 'test_token'
  },
  chatReducer: {
    messages: [],
    chats: []
  }
}
const dispatch = jest.fn()
let navigation
let store

beforeEach(() => {
  store = mockStore(initialState)
  navigation = { navigate: jest.fn() }
})

it('renders the neighbors list', async () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Neighbors navigation={navigation} dispatch={dispatch} />
    </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders empty chat', async () => {
  navigation.state = {
    params: {
      user: {
        id: 1,
        name: 'test',
        last_name: 'neighbor',
        apartments: [
          { id: 1, number: '1' }
        ]
      }
    }
  }
  const tree = renderer.create(
    <Provider store={store}>
      <Chat navigation={navigation} dispatch={dispatch} />
    </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders non-empty chat', async () => {
  navigation.state = {
    params: {
      chatId: 'test'
    }
  }
  store = mockStore({ ...initialState, chatReducer: {
    messages: {
      test: [
        {
          _id: 'test_1',
          text: 'Hola',
          createdAt: new Date('01/01/2018'),
          user: {
            name: 'test neighbor',
            _id: 1
          }
        }
      ]
    }
  } })
  const tree = renderer.create(
    <Provider store={store}>
      <Chat navigation={navigation} dispatch={dispatch} />
    </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

// it('stacks messages', async () => {
//   navigation.state = {
//     params: {
//       chatId: 'test'
//     }
//   };
//   store = mockStore({...initialState, chatReducer: {
//     messages: {
//       test: []
//     }
//   }});
//   let chat = renderer.create(
//     <Provider store={store}>
//       <Chat navigation={navigation} dispatch={dispatch} />
//     </Provider>
//   ).getInstance();
//   console.log(chat)
//   chat.send([{
//     _id: 'test_1',
//     text: 'test 1',
//     createdAt: new Date('01/01/2018'),
//     user: {
//       name: 'tester',
//       _id: 2
//     }
//   }]);
//   chat.send([{
//     _id: 'test_2',
//     text: 'test 2',
//     createdAt: new Date('01/01/2018'),
//     user: {
//       name: 'tester',
//       _id: 2
//     }
//   }]);
//   expect(store.messages.test.length).toBe(2);
// });


export function findById(tree, testID) {
  if (tree.props && tree.props.testID === testID) {
    return tree
  }
  if (tree.children && tree.children.length > 0) {
    let childs = tree.children
    for (let i = 0; i < childs.length; i++) {
      let item = findById(childs[i], testID)
      if (typeof (item) !== 'undefined') {
        return item
      }
    }
  }
  return null
}
