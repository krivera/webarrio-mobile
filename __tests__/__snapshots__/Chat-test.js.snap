// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty chat 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`renders non-empty chat 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`renders the neighbors list 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomColor": "#92a2a2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "padding": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Buscar"
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "apartments": Array [
            Object {
              "id": 1,
              "number": "1",
            },
          ],
          "id": 1,
          "last_name": "neighbor",
          "name": "test",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomColor": "#92a2a2",
              "borderBottomWidth": 0.5,
              "flexDirection": "row",
              "opacity": 1,
              "padding": 10,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={undefined}
          >
            <View
              style={
                Array [
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "#92a2a2",
                    "justifyContent": "center",
                    "marginRight": 5,
                  },
                  Object {
                    "borderRadius": 20,
                    "height": 40,
                    "width": 40,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "bottom": 0,
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "white",
                      },
                      Object {
                        "fontSize": 30,
                      },
                    ]
                  }
                >
                  T
                </Text>
              </View>
              <Image
                onLoad={[Function]}
                source={
                  Object {
                    "uri": undefined,
                  }
                }
                style={
                  Object {
                    "borderRadius": 20,
                    "height": 40,
                    "width": 40,
                  }
                }
              />
            </View>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            test
             de 
            1
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
